### PR TITLE
Move API system tests to cypress

### PR DESCRIPTION
**Pull request by laoneo**
Moves the Joomla API tests to cypress and removes the whole codeception dependency. There is only one remaining system tests for the back end media manager. As these tests are not executed in drone they will be ported in a second step.

Beside that it renames the tests/cypress folder to tests/System to be more inline with Unit and Integration.

### CODE REVIEW

**Comment by laoneo** [comment]
@dgrammatiko can you have a look here if the lint:testjs command can somehow be merged into the lint:js command? Thanks.

**Comment by dgrammatiko** [comment]
You could add another line that fires the existing tasks ie:

```json
   ""lint:js": "npm run lints-src && lint:testjs",
   "lint:js-src": "eslint --config build/.eslintrc --ignore-pattern '/media/' --ext .es6.js,.es6,.vue,.cy.js .",
     "lint:testjs": "eslint --config build/.eslintrc --ext .js tests/cypress",
```

or combine these as one task, ie:

```json
"lint:js": "eslint --config build/.eslintrc --ignore-pattern '/media/' --ext .es6.js,.es6,.vue,.cy.js . && eslint --config build/.eslintrc --ext .js tests/cypress",
```

File: tests/README.md [hunk 1]
```diff
@@ -15,6 +15,6 @@ Javascript tests
 ==========
 The javascript tests test the Joomla-specific Javascript code. For further information on the tests and on how to run them, please check out https://docs.joomla.org/Special:MyLanguage/Running_JavaScript_Tests_for_the_Joomla_CMS
 
-Codeception tests
+Cypress tests
 ==========
-The Codeception tests test the user interface in a real browser and the webservices API of Joomla with the help of Codeception. For further information on the tests and on how to run them, please check out [tests/Codeception/README.md](Codeception/README.md).
+The cypress tests test the user interface in a real browser and the webservices API of Joomla with the help of cypress.io. For further information on the tests and on how to run them, please check out [tests/cypress/README.md](cypress/README.md).
```
Comment (brianteeman) [inline tests/README.md]:
```suggestion
The Cypress tests test the user interface in a real browser and the webservices API of Joomla with the help of cypress.io. For further information on the tests and on how to run them, please check out [tests/cypress/README.md](cypress/README.md).
```

File: tests/cypress/integration/api/com_media/Files.cy.js [hunk 2]
```diff
@@ -0,0 +1,191 @@
...
+  });
+
+  it('can deliver a list of files in a subfolder', () => {
+    cy.api_get('/media/files/sampledata/cassiopeia/')
+      .then((response) => cy.wrap(response).its('body').its('data.0').its('attributes')
+        .its('name')
+        .should('include', 'nasa1-1200.jpg'));
+  });
+
+  it('can deliver a list of files wih an adapter', () => {
```
Comment (brianteeman) [inline tests/cypress/integration/api/com_media/Files.cy.js]:
```suggestion
  it('can deliver a list of files with an adapter', () => {
```

File: tests/cypress/integration/api/com_media/Files.cy.js [hunk 3]
```diff
@@ -0,0 +1,191 @@
...
+    cy.api_get('/media/files/local-images:/sampledata/cassiopeia/')
+      .then((response) => cy.wrap(response).its('body').its('data.0').its('attributes')
+        .its('name')
+        .should('include', 'nasa1-1200.jpg'));
+  });
+
+  it('can search in files', () => {
```
Comment (brianteeman) [inline tests/cypress/integration/api/com_media/Files.cy.js]:
```suggestion
  it('can search in filenames', () => {
```

File: tests/cypress/support/commands/api.js [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+  if (id.length > 0) {
+    return 'c2hhMjU2OjM6ZTJmMjJlYTNlNTU0NmM1MDJhYTIzYzMwN2MxYzAwZTQ5NzJhMWRmOTUyNjY5MTk2YjE5ODJmZWMwZTcxNzgwMQ==';
+  }
+
+  return cy.db_createUser({
+    id: 3,
+    name: 'API',
+    email: 'api@example.com',
+    username: 'api',
+    password: '123',
```
Comment (brianteeman) [inline tests/cypress/support/commands/api.js]: does this work? The default is for their to be a minimum password length of 8 characters

File: tests/cypress/drone-system-run.sh [hunk 5]
```diff
@@ -19,11 +19,12 @@ apache2ctl -D FOREGROUND &
...
-# cd /tests/www/$TEST_GROUP
 
-#export CYPRESS_CACHE_FOLDER=/tests/www/$DB_ENGINE/.cache
+cd /tests/www/$TEST_SUITE
+php installation/joomla.php install --verbose --site-name="Joomla CMS test" --admin-email=admin@example.org --admin-username=ci-admin --admin-user="jane doe" --admin-password=joomla-17082005 --db-type=$DB_ENGINE --db-host=$DB_HOST --db-name=test_joomla --db-pass=joomla_ut --db-user=root --db-encryption=0 --db-prefix=$TEST_GROUP
 
-#npx cypress install
-#npx cypress verify
-npx cypress run --browser=firefox --e2e --env db_type=$DB_ENGINE,db_host=$DB_HOST,db_password=joomla_ut,db_prefix="${TEST_GROUP}_" --config baseUrl=http://localhost/$TEST_GROUP,screenshotsFolder=$JOOMLA_BASE/tests/cypress/output/screenshots
+# If you have found this line failing on OSX you need to brew install gnu-sed like we mentioned in the codeception readme!
```
Comment (brianteeman) [inline tests/cypress/drone-system-run.sh]:
```suggestion
# If this line fails when using OSX then you need to brew install gnu-sed as stated in the Cypress readme!
```

Gammar changes but also I assume you meant cypress not codeception?

Comment (laoneo) [reply tests/cypress/drone-system-run.sh]: Thanks. I still try out thing, so this is kinda work in progress.

File: tests/cypress/drone-system-run.sh [hunk 6]
```diff
@@ -19,11 +19,13 @@ apache2ctl -D FOREGROUND &
...
-#export CYPRESS_CACHE_FOLDER=/tests/www/$DB_ENGINE/.cache
+cd /tests/www/$TEST_GROUP
+php installation/joomla.php install --verbose --site-name="Joomla CMS test" --admin-email=admin@example.org --admin-username=ci-admin --admin-user="jane doe" --admin-password=joomla-17082005 --db-type=$DB_ENGINE --db-host=$DB_HOST --db-name=test_joomla --db-pass=joomla_ut --db-user=root --db-encryption=0 --db-prefix="${TEST_GROUP}_"
 
-#npx cypress install
-#npx cypress verify
-npx cypress run --browser=firefox --e2e --env db_type=$DB_ENGINE,db_host=$DB_HOST,db_password=joomla_ut,db_prefix="${TEST_GROUP}_" --config baseUrl=http://localhost/$TEST_GROUP,screenshotsFolder=$JOOMLA_BASE/tests/cypress/output/screenshots
+# If you have found this line failing on OSX you need to brew install gnu-sed like we mentioned in the codeception readme!
+# This replaces the site secret in configuration.php so we can guarantee a consistent API token for our super user.
+sed -i "/\$secret/c\	public \$secret = 'tEstValue';" /tests/www/$TEST_SUITE/configuration.php
```
Comment (laoneo) [inline tests/cypress/drone-system-run.sh]:
```suggestion
sed -i "/\$secret/c\	public \$secret = 'tEstValue';" configuration.php
```

File: tests/cypress/support/commands/api.js (hunk 4 above)
Comment (laoneo) [reply tests/cypress/support/commands/api.js]: This is only needed for API authentication where we work with a token and no password.

File: tests/cypress/drone-system-run.sh [hunk 7]
```diff
@@ -14,6 +14,9 @@ echo "[RUNNER] Copy files to test installation"
 rsync -a --exclude-from=tests/cypress/exclude.txt $JOOMLA_BASE/ /tests/www/$TEST_GROUP/
 chown -R www-data /tests/www/$TEST_GROUP/
 
+# Is needed for media manager tests
```
Comment (brianteeman) [inline tests/cypress/drone-system-run.sh]:
```suggestion
# Required for media manager tests
```

File: libraries/src/Helper/MediaHelper.php [hunk 8]
```diff
@@ -146,7 +146,7 @@ private function checkMimeType($mime, $component = 'com_media'): bool
             );
 
             // Get the mime type configuration
-            $allowedMime = array_map('trim', explode(',', $allowedMime));
+            $allowedMime = array_map('trim', explode(',', str_replace('\\', '', $allowedMime)));
```
Comment (wilsonge) [inline libraries/src/Helper/MediaHelper.php:149]: I'm sorry if you've had to change the code here as part of this PR it needs properly testing. Merging stuff without any changes when you're changing library code is wrong

**Pull request merged**